src/operations/blas2/spmv.hpp

```diff
@@ -0,0 +1,284 @@
+ *
+ **************************************************************************/
+
+#ifndef SPMV_HPP
+#define SPMV_HPP
+#include "operations/blas2_trees.h"
+#include "operations/blas_operators.hpp"
+#include "views/view_sycl.hpp"
+#include <stdexcept>
+#include <vector>
+namespace blas {
+
+/**
+ * @struct Spmv
+ * @brief Tree node representing a symmetric band matrix_ vector_
+ * multiplication.
+ */
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range_x, uint32_t local_range_y, bool is_upper>
+SYCL_BLAS_INLINE
+Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y, is_upper>::Spmv(
+    lhs_t &_l, matrix_t &_matrix, vector_t &_vector,
+    typename Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y,
+                  is_upper>::value_t _alpha,
+    typename Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y,
+                  is_upper>::value_t _beta)
+    : lhs_(_l),
+      matrix_(_matrix),
+      vector_(_vector),
+      alpha_(_alpha),
+      beta_(_beta) {}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range_x, uint32_t local_range_y, bool is_upper>
+SYCL_BLAS_INLINE typename Spmv<lhs_t, matrix_t, vector_t, local_range_x,
+                               local_range_y, is_upper>::index_t
+Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y,
+     is_upper>::get_size() const {
+  return matrix_.get_size();
+}
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range_x, uint32_t local_range_y, bool is_upper>
+SYCL_BLAS_INLINE bool
+Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y,
+     is_upper>::valid_thread(cl::sycl::nd_item<1> ndItem) const {
+  // Valid threads are established by ::eval.
+  return true;
+}
+
+template <typename lhs_t, typename matrix_t, typename vector_t,
+          uint32_t local_range_x, uint32_t local_range_y, bool is_upper>
+template <typename sharedT>
+SYCL_BLAS_INLINE typename Spmv<lhs_t, matrix_t, vector_t, local_range_x,
+                               local_range_y, is_upper>::value_t
+Spmv<lhs_t, matrix_t, vector_t, local_range_x, local_range_y, is_upper>::eval(
+    sharedT shrMem, cl::sycl::nd_item<1> ndItem) {
+  const index_t gid = ndItem.get_group(0);
+
+  constexpr index_t x_range = local_range_x;
+  constexpr index_t y_range = local_range_y;
+  constexpr index_t y_chunck = x_range / y_range;
+  constexpr index_t loc_lda = x_range + 1;
+
+  const index_t l_idx = ndItem.get_local_id(0) % x_range;
+  const index_t l_idy = ndItem.get_local_id(0) / x_range;
+  const index_t l_y_offset = y_chunck * l_idy;
+
+  const index_t N = lhs_.get_size();
+  const index_t nblock = (N + x_range - 1) / x_range;
+
+  value_t *const loc_x = shrMem.localAcc.get_pointer();
+  value_t *const loc_A = loc_x + x_range;
+
+  // ------------------------------------------------------------------------ //
+
+  auto _mat_J_offset = [&N](const index_t &_J) {
```

Comment:
I think it's better to use `if` conditions instead of these `lambda` functions, the compiler may be able to flatten the code out the `if` condition since the code is dependent on a template parameter but the lambda object may not go away.

Reply:
Maybe we could use `if constexpr` to be sure the compiler selects the correct one, couldn't we?

Reply:
I introduced these lambda functions for the sake of making this kernel more readable/maintainable by avoiding the code duplication, and potentially (one day with a similar approach) share the common code among different kernels.

So far, these lambdas are handled correctly by the compiler. They are inlined automatically and the ternary operator that they contain is simplified at compile time since depends on a constant. This basic optimization is performed easily by clang. I understand your concerns related to the possibility that those lambdas could not be inlined, but for now the inline issues we encountered are related to functions having many instructions/arguments.